flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java

```diff
@@ -0,0 +1,366 @@
+		flinkKubeClient.createTaskManagerPod(new TaskManagerPodParameter(
+			previewPodName,
+			new ArrayList<>(),
+			1024,
+			1,
+			new HashMap<>()));
+		final KubernetesClient client = getKubeClient();
+		assertEquals(1, client.pods().list().getItems().size());
+
+		final String taskManagerPrefix = CLUSTER_ID + "-taskmanager-";
+		final TestingKubernetesResourceManager resourceManager = this.createAndStartResourceManager(flinkConfig);
+		// Register the previous taskmanager, no new pod should be created
+		registerTaskExecutor(resourceManager, new ResourceID(previewPodName));
+		registerSlotRequest(resourceManager);
+		assertEquals(1, client.pods().list().getItems().size());
+
+		// Register a new slot request, a new taskmanger pod will be created with attempt2
+		registerSlotRequest(resourceManager);
+		assertEquals(2, client.pods().list().getItems().size());
+		assertThat(client.pods().list().getItems().stream()
+				.map(e -> e.getMetadata().getName())
+				.collect(Collectors.toList()),
+			Matchers.containsInAnyOrder(taskManagerPrefix + "1-1", taskManagerPrefix + "2-1"));
+
+		resourceManager.getRpcService().stopService().get();
+	}
+
+	private TestingKubernetesResourceManager createAndStartResourceManager(Configuration configuration) throws Exception {
+
+		final TestingRpcService rpcService = new TestingRpcService(configuration);
+		final TestingFatalErrorHandler testingFatalErrorHandler = new TestingFatalErrorHandler();
+		org.apache.flink.runtime.resourcemanager.utils.MockResourceManagerRuntimeServices rmServices =
+			new org.apache.flink.runtime.resourcemanager.utils.MockResourceManagerRuntimeServices(rpcService, TIMEOUT);
+
+		final TestingKubernetesResourceManager kubernetesResourceManager = new TestingKubernetesResourceManager(
+			rpcService,
+			"kubernetesResourceManager",
+			ResourceID.generate(),
+			configuration,
+			rmServices.highAvailabilityServices,
+			rmServices.heartbeatServices,
+			rmServices.slotManager,
+			rmServices.jobLeaderIdService,
+			new ClusterInformation("localhost", 1234),
+			testingFatalErrorHandler,
+			UnregisteredMetricGroups.createUnregisteredResourceManagerMetricGroup()
+		);
+		kubernetesResourceManager.start();
+		rmServices.grantLeadership();
+		return kubernetesResourceManager;
+	}
+
+	private void registerSlotRequest(TestingKubernetesResourceManager resourceManager) throws Exception {
+
+		CompletableFuture<?> registerSlotRequestFuture = resourceManager.runInMainThread(() -> {
+			resourceManager.getSlotManager().registerSlotRequest(
+				new SlotRequest(new JobID(), new AllocationID(), ResourceProfile.UNKNOWN, jobManagerHost));
+			return null;
+		});
+		registerSlotRequestFuture.get();
+	}
+
+	private void registerTaskExecutor(
+			TestingKubernetesResourceManager resourceManager,
+			ResourceID resourceID) throws Exception {
+		// Remote task executor registers with KubernetesResourceManager.
+		final TaskExecutorGateway mockTaskExecutorGateway = mock(TaskExecutorGateway.class);
```

Comment:
it is better to not mock any components in tests.

Reply:
We should mock `mockTaskExecutorGateway` here. Because there is no really `TaskManager` running there.

Reply:
I guess there are other ways to work around to void touching `TaskExecutorGateway` in the tests. The mock way is actually not suggested by ground truth. I would further consider it.

Reply:
I will replace the `mockTaskExecutorGateway` to `TestingTaskExecutorGateway` to eliminate the mock components.